### PR TITLE
fix: Fixes crash when voice input is closed

### PR DESCRIPTION
**Pull request by chiragw15**
Fixes issue #657 

Changes: Added a check to see if recognizer is null.

### CODE REVIEW

**Comment by iamareebjamal** [comment]
I'll pull and see if this fixed the issue

**Comment by iamareebjamal** [comment]
This did not fix it. But I digged into the error. The problem was that my emulator didn't have speech input system, so the SpeechRecognizer was logging error that
```
E/SpeechRecognizer: no selected voice recognition service
```

So, I opened the class and saw that it also notifies the listener with `onError` callback : 
````Java
if (mServiceComponent == null) {
    String serviceComponent = Settings.Secure.getString(mContext.getContentResolver(),
            Settings.Secure.VOICE_RECOGNITION_SERVICE);
    
    if (TextUtils.isEmpty(serviceComponent)) {
        Log.e(TAG, "no selected voice recognition service");
        mListener.onError(ERROR_CLIENT);
        return;
    }
    
    serviceIntent.setComponent(ComponentName.unflattenFromString(serviceComponent));                
} else {
    serviceIntent.setComponent(mServiceComponent);
}
````

But the listener isn't `RecognitionListener`, it is some internal listener which then sends the message to Recognition listener. But I checked and onError is never called for some reason.

I also noticed that there is a GC call right after I press the mic option, so if they are using WeakReference, then listener can be dereferenced at that time and become null.

I tried thinking of a solution but when the callbacks aren't working, there is nothing much we can do except to pull the internal code outside to check if there are no speech input methods and not try to start listening if this is the case

**Comment by chiragw15** [comment]
So, what do you suggest? I guess, this problem of not having a speech input method should not occur on a mobile device. I had tested this for 2-3 mobile phones, worked fine for them. So, what should we do? `except to pull the internal code outside to check if there are no speech input methods and not try to start listening if this is the case` Should we implement this?

**Comment by iamareebjamal** [comment]
The thing is, application should not crash in any condition. A simple check won't hurt

**Comment by chiragw15** [comment]
@iamareebjamal can you tell me which emulator you are using? I have done some changes. Needed to test them.

**Comment by iamareebjamal** [comment]
Genymotion

**Comment by chiragw15** [comment]
@iamareebjamal review please.

File: app/src/main/java/org/fossasia/susi/ai/activities/MainActivity.java [hunk 1]
```diff
@@ -405,8 +409,13 @@ protected void onCreate(Bundle savedInstanceState) {
 
         if(ActivityCompat.checkSelfPermission(this,
                 Manifest.permission.RECORD_AUDIO) == PackageManager.PERMISSION_GRANTED) {
-            micCheck = true;
-            PrefManager.putBoolean(Constant.MIC_INPUT, true);
+            if(MediaUtil.isAvailableForVoiceInput(MainActivity.this)) {
```
Comment (iamareebjamal) [inline app/src/main/java/org/fossasia/susi/ai/activities/MainActivity.java]: Store in a boolean `voiceInputAvailable` and 

````Java
micCheck = voiceInputAvailable;
PrefManager.putBoolean(Constant.MIC_INPUT, voiceInputAvailable);
````

File: app/src/main/java/org/fossasia/susi/ai/activities/MainActivity.java [hunk 2]
```diff
@@ -726,8 +735,13 @@ public void onRequestPermissionsResult(int requestCode, String permissions[], in
                     micCheck = false;
                     PrefManager.putBoolean(Constant.MIC_INPUT, false);
                 } else {
-                    micCheck = true;
-                    PrefManager.putBoolean(Constant.MIC_INPUT, true);
+                    if(MediaUtil.isAvailableForVoiceInput(MainActivity.this)) {
```
Comment (iamareebjamal) [inline app/src/main/java/org/fossasia/susi/ai/activities/MainActivity.java]: Same here

File: app/src/main/java/org/fossasia/susi/ai/activities/MainActivity.java [hunk 3]
```diff
@@ -752,8 +766,13 @@ public void onRequestPermissionsResult(int requestCode, String permissions[], in
...
```
Comment (iamareebjamal) [inline app/src/main/java/org/fossasia/susi/ai/activities/MainActivity.java]: Same here, why not create a function?

File: app/src/main/java/org/fossasia/susi/ai/activities/SettingsActivity.java [hunk 4]
```diff
@@ -84,6 +89,20 @@ public void execute(Realm realm) {
                     return true;
                 }
             });
+
+            micSettings = getPreferenceManager().findPreference("Mic_input");
+            if(ActivityCompat.checkSelfPermission(getContext(),
+                    Manifest.permission.RECORD_AUDIO) == PackageManager.PERMISSION_GRANTED) {
+                if(MediaUtil.isAvailableForVoiceInput(getContext())) {
```
Comment (iamareebjamal) [inline app/src/main/java/org/fossasia/susi/ai/activities/SettingsActivity.java]: Same here

File: app/src/main/java/org/fossasia/susi/ai/helper/MediaUtil.java [hunk 5]
```diff
@@ -0,0 +1,54 @@
...
+        MediaRecorder recorder = new MediaRecorder();
+        recorder.setAudioSource(MediaRecorder.AudioSource.MIC);
+        recorder.setOutputFormat(MediaRecorder.OutputFormat.DEFAULT);
+        recorder.setAudioEncoder(MediaRecorder.AudioEncoder.DEFAULT);
+        recorder.setOutputFile(new File(context.getCacheDir(), "MediaUtil#micAvailTestFile").getAbsolutePath());
+        boolean available = true;
+        try {
+            recorder.prepare();
+        }
+        catch (IOException exception) {
```
Comment (iamareebjamal) [inline app/src/main/java/org/fossasia/susi/ai/helper/MediaUtil.java]: Move catch up

**Comment by chiragw15** [comment]
@iamareebjamal done.

**Pull request merged**